### PR TITLE
[Lib] Vector and DataPoint classes for machine learning in general

### PR DESCRIPTION
**Pull request by Kelvin-Ng**
Issue #106

### CODE REVIEW

**Comment by kygx-legend** [comment]
@ddmbr @Yuzhen11 @zzxx-husky any comment?

File: lib/vector.hpp [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+        for (auto entry : b) {
+            vec[entry.fea] = entry.val;
+        }
+    }
+
+    Vector<T, false>(int feature_num, T val)
+        : feature_num(feature_num),
+          vec(feature_num, val) {}
+
+    Vector<T, false>(int feature_num, T& val)
```
Comment (Kelvin-Ng) [inline lib/vector.hpp]: The line numbers seems outdated. Could you please mark the line number again?

File: lib/vector.hpp [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+    }
+
+    // TODO: how to make the following function available?
+    /*inline void set(int idx, const T& val) {
+        vec[idx] = val;
+    }*/
+
+    inline void set(int idx, T val) {
+        vec[idx] = std::move(val);
+    }
```
Comment (Kelvin-Ng) [inline lib/vector.hpp]: Yes, you are right!

File: lib/vector.hpp [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+
+    Vector<T, false>(const SparseVector<T>& b) : feature_num(b.get_feature_num()), vec(b.get_feature_num()) {
+        for (auto entry : b) {
+            vec[entry.fea] = entry.val;
+        }
+    }
+
+    Vector<T, false>(int feature_num, T val) : feature_num(feature_num), vec(feature_num, val) {}
+
+    Vector<T, false>(int feature_num, T& val) : feature_num(feature_num), vec(feature_num, val) {}
```
Comment (Yuzhen11) [inline lib/vector.hpp]: I mean these two. Line 137 and Line 139. I don't get it.

Comment (Kelvin-Ng) [inline lib/vector.hpp]: I see. Line 137 should be T&& value

File: lib/vector.hpp [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+    FeaValPair(int fea, T val) : fea(fea), val(val) {}
+    // TODO: I want fea to be a const, but this will make operator=() unavailable and break the use of STL containers
+    int fea;
+    T val;
+};
+
+template<typename T>
+class Vector<T, false> {
+public:
+    typedef typename std::vector<T>::iterator Iterator;
```
Comment (Kelvin-Ng) [inline lib/vector.hpp]: I used `using` at first because only `using` supports template.

**Comment by Kelvin-Ng** [comment]
By the way, here is an example of why the iterators are useful: https://github.com/Kelvin-Ng/VR-SGD-Implementation/blob/master/lib/prox.hpp (`prox_l1()`).

File: lib/vector.hpp [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+            return it;
+        }
+
+        FeaValIterator operator--(int) {
+            FeaValIterator it(*this);
+            idx--;
+            return it;
+        }
+
+        bool operator==(const FeaValIterator& b) const { return idx == b.idx; }
```
Comment (zzxx-husky) [inline lib/vector.hpp]: Don't quite understand why only the member `idx` is compared and `vec` is not.

**Comment by zzxx-husky** [comment]
LGTM~

File: lib/vector.hpp (hunk 5 above)
Comment (Kelvin-Ng) [inline lib/vector.hpp]: My fault. You are right.

Comment (Kelvin-Ng) [inline lib/vector.hpp]: Should compare the address.

Comment (Kelvin-Ng) [inline lib/vector.hpp]: Fixed

**Comment by Kelvin-Ng** [comment]
@ddmbr You will be killed by @kygx-legend because you merge it without unit tests.

**Comment by kygx-legend** [comment]
Implement the tests next PR. :)

**Pull request merged**